lib/kafka_gen_stage/consumer_logic.ex

```diff
@@ -83,7 +67,7 @@ defmodule KafkaGenStage.ConsumerLogic do
           high_wm :: non_neg_integer()
         ) :: dispatch()
   defp prepare_dispatch(queue, 0, to_send, to_ack, _high_wm) do
-    {to_send, to_ack, 0, queue}
+    {to_send, to_ack, queue}
```

Comment:
Now that we've simplified the function it becomes obvious it's not really a `prepare_dispatch`, all it does it dequeue as many as possible up to `demand` msgs from the queue. The `high_wm` parameter isn't used at all. `to_ack` is simply the offset of the last message dequeued, which conveniently happens to be the first in the returned list, thus cheap to read.

Therefore I suggest:

* turn this into a `dequeue` function that takes up to `demand` msgs from `queue`
* in `prepare_dispatch/4` calculate to_ack as `case to_send do [] -> :no_ack, [{offset, …}|_] -> offset end`
* drop the `dispatch()` type
* rename `return_dispatch()` to `dispatch()`

Reply:
I thought about `dequeue` too. I will do it.